pkg/elle/core/depend.go

```diff
@@ -0,0 +1,212 @@
+	}
+	res := result.(RealtimeDependent)
+	s := fmt.Sprintf("%s complete at index %d, ", preName, res.postStart.Index)
+
+	if !res.postStart.Time.IsZero() && !res.prefixEnd.Time.IsZero() {
+		t1, t2 := res.prefixEnd.Time, res.postStart.Time
+		if t1.Before(t2) {
+			delta := t2.Sub(t1)
+			s += fmt.Sprintf("%v seconds just ", delta.Milliseconds())
+		}
+	}
+
+	s += fmt.Sprintf("before the invocation of %s at index %d", postName, res.prefixEnd.Index)
+	return s
+}
+
+type RealtimeDependent struct {
+	prefixEnd *Op
+	postStart *Op
+}
+
+func (_ RealtimeDependent) Type() DependType {
+	return RealtimeDepend
+}
+
+// Note: MonotonicKey is used on rw_register, so I don't explain it currently.
+// MonotonicKeyExplainer ...
+type MonotonicKeyExplainer struct{}
+
+func (e MonotonicKeyExplainer) ExplainPairData(p1, p2 PathType) ExplainResult {
+	panic("implement me")
+}
+
+// RenderExplanation render explanation
+func (e MonotonicKeyExplainer) RenderExplanation(result ExplainResult, preName, postName string) string {
+	panic("impl me")
+}
+
+type ICycleExplainer interface {
+	ExplainCycle(pairExplainer DataExplainer, circle Circle) (Circle, []Step)
+	RenderCycleExplanation(explainer DataExplainer, circle Circle, steps []Step) string
+}
+
+// TODO: add panic & recover logic here.
+func explainCyclePairData(pairExplainer DataExplainer, p1, p2 PathType) ExplainResult {
+	return pairExplainer.ExplainPairData(p1, p2)
+}
+
+// CycleExplainer provides the step-by-step explanation of the relationships between pairs of operations
+type CycleExplainer struct{}
+
+// Explain for the whole scc.
+func (c *CycleExplainer) ExplainCycle(explainer DataExplainer, circle Circle) (Circle, []Step) {
+	var steps []Step
+	for i := 1; i < len(circle.Path); i++ {
+		res := explainer.ExplainPairData(circle.Path[i-1], circle.Path[i])
+		steps = append(steps, Step{Result: res})
+	}
+	return circle, steps
+}
+
+type OpBinding struct {
+	Operation Op
+	Name      string
+}
+
+func (c *CycleExplainer) RenderCycleExplanation(explainer DataExplainer, circle Circle, steps []Step) string {
+	var bindings []OpBinding
+	for i, v := range circle.Path {
+		bindings = append(bindings, OpBinding{
+			Operation: v,
+			Name:      fmt.Sprintf("T%d", i),
+		})
+	}
+	bindingsExplain := explainBindings(bindings)
+	stepsResult := explainCycleOps(explainer, bindings, steps)
+	return bindingsExplain + "\n" + stepsResult
+}
+
+// Takes a seq of [name op] pairs, and constructs a string naming each op.
+func explainBindings(bindings []OpBinding) string {
+	var seq []string
+	seq = []string{"Let:"}
+	for _, v := range bindings {
+		seq = append(seq, fmt.Sprintf(" %s = %s", v.Name, v.Operation.String()))
+	}
+	return strings.Join(seq, "\n")
+}
+
+func explainCycleOps(pairExplainer DataExplainer, bindings []OpBinding, steps []Step) string {
+	var explainitions []string
+	for i := 1; i < len(bindings); i++ {
+		res := pairExplainer.ExplainPairData(bindings[i-1].Operation, bindings[i].Operation)
```

Comment:
don't need explain again, we can reuse steps here, it's the result of explaination